src/Plugin/Field/FieldWidget/ProductVariationOptionWidgetBase.php

```diff
@@ -0,0 +1,191 @@
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Field\FieldDefinitionInterface $field_definition
+   *   The definition of the field to which the widget is associated.
+   * @param array $settings
+   *   The widget settings.
+   * @param array $third_party_settings
+   *   Any third party settings.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\Core\Entity\EntityRepositoryInterface $entity_repository
+   *   The entity repository.
+   * @param \Drupal\Core\TempStore\PrivateTempStoreFactory $temp_store_factory
+   *   The tempstore factory.
+   */
+  public function __construct($plugin_id, $plugin_definition, FieldDefinitionInterface $field_definition, array $settings, array $third_party_settings, EntityTypeManagerInterface $entity_type_manager, EntityRepositoryInterface $entity_repository, PrivateTempStoreFactory $temp_store_factory) {
+    parent::__construct($plugin_id, $plugin_definition, $field_definition, $settings, $third_party_settings);
+
+    $this->entityRepository = $entity_repository;
+    $this->variationStorage = $entity_type_manager->getStorage('commerce_product_variation');
+    $this->tempStore = $temp_store_factory;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $plugin_id,
+      $plugin_definition,
+      $configuration['field_definition'],
+      $configuration['settings'],
+      $configuration['third_party_settings'],
+      $container->get('entity_type.manager'),
+      $container->get('entity.repository'),
+      $container->get('tempstore.private')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function isApplicable(FieldDefinitionInterface $field_definition) {
+    $entity_type = $field_definition->getTargetEntityTypeId();
+    $field_name = $field_definition->getName();
+    return $entity_type == 'commerce_order_item' && $field_name == 'purchased_entity';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
+    // Assumes that the variation ID comes from an $element['variation'] built
+    // in formElement().
+    foreach ($values as $key => $value) {
+      $this->tempStore->get('commerce_option')->set('options', $value['options']);
+      $values[$key] = [
+        'target_id' => $value['variation'],
+      ];
+    }
+
+    return $values;
+  }
+
+  /**
+   * #ajax callback: Replaces the rendered fields on variation change.
+   *
+   * Assumes the existence of a 'selected_variation' in $form_state.
+   */
+  public static function ajaxRefresh(array $form, FormStateInterface $form_state) {
+    /** @var \Drupal\Core\Render\MainContent\MainContentRendererInterface $ajax_renderer */
+    $ajax_renderer = \Drupal::service('main_content_renderer.ajax');
```

Comment:
Couldn't we inject all this in via dependency injection?

Reply:
Not really, since we are in the static context and $this is not available

Reply:
Ah yeah.